packages/xo-server/src/xo-mixins/metadata-backups.js

```diff
@@ -0,0 +1,246 @@
+      app.registerJobExecutor(
+        METADATA_BACKUP_JOB_TYPE,
+        this._executor.bind(this)
+      )
+    })
+  }
+
+  @deferrable
+  async _executor($defer, { cancelToken, job: job_, schedule }): Executor {
+    if (schedule === undefined) {
+      throw new Error('backup job cannot run without a schedule')
+    }
+
+    const job: MetadataBackupJob = (job_: any)
+    const remoteIds = unboxIdsFromPattern(job.remotes)
+    if (remoteIds.length === 0) {
+      throw new Error('metadata backup job cannot run without remotes')
+    }
+
+    const poolIds = unboxIdsFromPattern(job.pools)
+    const isEmptyPools = poolIds.length === 0
+    if (!job.xoMetadata && isEmptyPools) {
+      throw new Error('no metadata mode found')
+    }
+
+    const app = this._app
+    const { retentionXoMetadata, retentionPoolMetadata } =
+      job?.settings[schedule.id] || {}
+
+    const scheduleDir = `${METADATA_BACKUP_DIR}/${schedule.id}`
+    const timestamp = safeDateFormat(Date.now())
+    const files = []
+    if (job.xoMetadata && retentionXoMetadata > 0) {
+      const xoDir = `${scheduleDir}/xo`
+      const dir = `${xoDir}/${timestamp}`
+
+      const data = JSON.stringify(await app.exportConfig(), null, 2)
+      const fileName = `${dir}/data.json`
+
+      const metadata = JSON.stringify(
+        {
+          jobId: job.id,
+          scheduleId: schedule.id,
+          timestamp,
+        },
+        null,
+        2
+      )
+      const metaDataFileName = `${dir}/metadata.json`
+
+      files.push({
+        executeBackup: async handler => {
+          await Promise.all([
+            handler.outputFile(fileName, data),
+            handler.outputFile(metaDataFileName, metadata),
+          ])
+          $defer.onFailure(() => handler.rmtree(dir))
+        },
+        dir: xoDir,
+        retention: retentionXoMetadata,
+      })
+    }
+    if (!isEmptyPools && retentionPoolMetadata > 0) {
+      files.push(
+        ...(await Promise.all(
+          poolIds.map(async id => {
+            const metadataDir = `${scheduleDir}/${id}`
+            const dir = `${metadataDir}/${timestamp}`
+
+            // TODO: export the metadata only once then split the stream between remotes
+            const stream = await app.getXapi(id).exportPoolMetadata(cancelToken)
+            const fileName = `${dir}/data`
+
+            const metadata = JSON.stringify(
+              {
+                jobId: job.id,
+                scheduleId: schedule.id,
+                poolId: id,
+                timestamp,
+              },
+              null,
+              2
+            )
+            const metaDataFileName = `${dir}/metadata.json`
+
+            return {
+              executeBackup: async handler => {
+                const [outputStream] = await Promise.all([
+                  handler.createOutputStream(fileName),
+                  handler.outputFile(metaDataFileName, metadata),
+                ])
+                pipeline(stream, outputStream)
```

Comment:
@julien-f  I tried to  use `promise-toolbox/promisify`  on the function `pipeline` but it not resolves. Did you have any idea for the reason? Thank you :)

Reply:
It should work, have you tested it with a basic script?

Reply:
Not yet, i will test it :+1: